src/msg_recorder/recorder.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include "cris/core/msg/message.h"
+#include "cris/core/msg/node.h"
+#include "cris/core/msg_recorder/record_file.h"
+
+#include <filesystem>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+
+namespace cris::core {
+
+class MessageRecorder : public CRNamedNode<MessageRecorder> {
+   public:
+    using Base = CRNamedNode<MessageRecorder>;
+
+    explicit MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner);
+
+    MessageRecorder(const MessageRecorder&) = delete;
+
+    MessageRecorder(MessageRecorder&&) = delete;
+
+    MessageRecorder& operator=(const MessageRecorder&) = delete;
+
+    ~MessageRecorder();
+
+    template<CRMessageType message_t>
+    void RegisterChannel(const channel_subid_t subid, const std::string& alias = "");
+
+    std::filesystem::path GetRecordDir() const;
+
+   private:
+    using msg_serializer = std::function<std::string(const CRMessageBasePtr&)>;
+
+    RecordFile* CreateFile(const std::string& message_type, const channel_subid_t subid, const std::string& alias);
+
+    static std::string RecordDirNameGenerator();
+
+    std::filesystem::path                    record_dir_;
```

Comment:
const?

Reply:
Why? I guess you may need a very strong reason to make a data member const.

Reply:
Because dir/fn are representing something on disk, i.e. it a pointer and can leak if modified.